asserts/extkeypairmgr.go

```diff
@@ -0,0 +1,260 @@
+	}
+	if err := em.keyMgr("features", nil, nil, &feats); err != nil {
+		return err
+	}
+	if !strutil.ListContains(feats.Signing, "RSA-PKCS") {
+		return fmt.Errorf("external keypair manager %q missing support for RSA-PKCS signing", em.keyMgrPath)
+	}
+	if !strutil.ListContains(feats.PublicKeys, "DER") {
+		return fmt.Errorf("external keypair manager %q missing support for public key DER output format", em.keyMgrPath)
+	}
+	return nil
+}
+
+func (em *ExternalKeypairManager) keyNames() ([]string, error) {
+	var knames struct {
+		Names []string `json:"key-names"`
+	}
+	if err := em.keyMgr("key-names", nil, nil, &knames); err != nil {
+		return nil, fmt.Errorf("cannot get all external keypair manager key names: %v", err)
+	}
+	return knames.Names, nil
+}
+
+func (em *ExternalKeypairManager) findByName(name string) (PublicKey, *rsa.PublicKey, error) {
+	var k []byte
+	err := em.keyMgr("get-public-key", []string{"-f", "DER", "-k", name}, nil, &k)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot find external key: %v", err)
+	}
+	pubk, err := x509.ParsePKIXPublicKey(k)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot decode external key %q: %v", name, err)
+	}
+	rsaPub, ok := pubk.(*rsa.PublicKey)
+	if !ok {
+		return nil, nil, fmt.Errorf("expected RSA public key, got instead: %T", pubk)
+	}
+	pubKey := RSAPublicKey(rsaPub)
+	return pubKey, rsaPub, nil
+}
+
+func (em *ExternalKeypairManager) Export(keyName string) ([]byte, error) {
+	pubKey, _, err := em.findByName(keyName)
+	if err != nil {
+		return nil, err
+	}
+	return EncodePublicKey(pubKey)
+}
+
+func (em *ExternalKeypairManager) loadKey(name string) (*cachedExtKey, error) {
+	id, ok := em.nameToID[name]
+	if ok {
+		return em.cache[id], nil
+	}
+	pubKey, rsaPub, err := em.findByName(name)
+	if err != nil {
+		return nil, err
+	}
+	id = pubKey.ID()
+	em.nameToID[name] = id
+	cachedKey := &cachedExtKey{
+		pubKey: pubKey,
+		signer: &extSigner{
+			keyName: name,
+			rsaPub:  rsaPub,
+			// signWith is filled later
+		},
+	}
+	em.cache[id] = cachedKey
+	return cachedKey, nil
+}
+
+func (em *ExternalKeypairManager) privateKey(cachedKey *cachedExtKey) PrivateKey {
+	if cachedKey.privKey == nil {
+		extSigner := cachedKey.signer
+		// fill signWith
+		extSigner.signWith = em.signWith
+		signer := packet.NewSignerPrivateKey(v1FixedTimestamp, extSigner)
+		signk := openpgpPrivateKey{privk: signer}
+		extKey := &extPGPPrivateKey{
+			pubKey:     cachedKey.pubKey,
+			from:       fmt.Sprintf("external keypair manager %q", em.keyMgrPath),
+			externalID: extSigner.keyName,
+			bitLen:     extSigner.rsaPub.N.BitLen(),
+			doSign:     signk.sign,
+		}
+		cachedKey.privKey = extKey
+	}
+	return cachedKey.privKey
+}
+
+func (em *ExternalKeypairManager) GetByName(keyName string) (PrivateKey, error) {
+	cachedKey, err := em.loadKey(keyName)
+	if err != nil {
+		return nil, err
+	}
+	return em.privateKey(cachedKey), nil
+}
+
+func (em *ExternalKeypairManager) Put(privKey PrivateKey) error {
```

Comment:
Is this used (or tested) at all?

Reply:
it's part of the interface but we don't use it from snap either for this or GPG, this implementation is copied from the GPG one